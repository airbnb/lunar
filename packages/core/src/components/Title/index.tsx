import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';

const stateProp = mutuallyExclusiveTrueProps('muted', 'inverted', 'primary');
const alignProp = mutuallyExclusiveTrueProps('centerAlign', 'endAlign');

export type Props = {
  /** Align the text in the center. */
  centerAlign?: boolean;
  /** The text to render. */
  children?: React.ReactNode;
  /** Align the text on the end. */
  endAlign?: boolean;
  /** Render the text inline instead of block. */
  inline?: boolean;
  /** Invert text colors. */
  inverted?: boolean;
  /** Heading level of the text, with 1 being the greatest. */
  level: 1 | 2 | 3;
  /** Mark the text as muted. */
  muted?: boolean;
  /** Render with primary color text. */
  primary?: boolean;
};

/** Display a string of text as a heading and or section title. */
export class Title extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    centerAlign: alignProp,
    endAlign: alignProp,
    inverted: stateProp,
    muted: stateProp,
    primary: stateProp,
  };

  static defaultProps = {
    centerAlign: false,
    children: null,
    endAlign: false,
    inline: false,
    inverted: false,
    muted: false,
    primary: false,
  };

  render() {
    const {
      centerAlign,
      children,
      endAlign,
      inline,
      inverted,
      level,
      muted,
      primary,
      styles,
    } = this.props;
    const Tag: 'h1' | 'h2' | 'h3' = `h${level}` as any;

    return (
      <Tag
        {...css(
          styles.title,
          level === 1 && styles.title_level1,
          level === 2 && styles.title_level2,
          level === 3 && styles.title_level3,
          inline && styles.title_inline,
          inverted && styles.title_inverted,
          muted && styles.title_muted,
          primary && styles.title_primary,
          centerAlign && styles.title_center,
          endAlign && styles.title_right,
        )}
      >
        {children}
      </Tag>
    );
  }
}

export default withStyles(({ color, font }) => ({
  title: {
    ...font.textReset,
    color: color.accent.text,
  },

  title_inline: {
    display: 'inline',
  },

  title_level1: {
    ...font.title1,
  },

  title_level2: {
    ...font.title2,
  },

  title_level3: {
    ...font.title3,
  },

  title_inverted: {
    color: color.base,
  },

  title_muted: {
    color: color.muted,
  },

  title_primary: {
    color: color.core.primary[3],
  },

  title_center: {
    textAlign: 'center',
  },

  title_right: {
    textAlign: 'right',
  },
}))(Title);
