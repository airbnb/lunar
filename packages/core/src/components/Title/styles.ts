import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, font }) => ({
  title: {
    ...font.textReset,
    color: color.accent.text,
  },

  title_inline: {
    display: 'inline',
  },

  title_level1: {
    ...font.title1,
  },

  title_level2: {
    ...font.title2,
  },

  title_level3: {
    ...font.title3,
  },

  title_inverted: {
    color: color.base,
  },

  title_muted: {
    color: color.muted,
  },

  title_primary: {
    color: color.core.primary[3],
  },

  title_center: {
    textAlign: 'center',
  },

  title_right: {
    textAlign: 'right',
  },
});

export default styleSheet;

export { styleSheet };
