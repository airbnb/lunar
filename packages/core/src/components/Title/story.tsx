import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Title from '.';

export default {
  title: 'Core/Title',
  parameters: {
    inspectComponents: [Title],
  },
};

export function titlesWithDifferentHeadingLevels() {
  return (
    <>
      <Title level={1}>
        <LoremIpsum short />
      </Title>

      <Title level={2}>
        <LoremIpsum short />
      </Title>

      <Title level={3}>
        <LoremIpsum short />
      </Title>
    </>
  );
}

titlesWithDifferentHeadingLevels.story = {
  name: 'Titles with different heading levels.',
};

export function withDifferentStatesMutedInvertedAndPrimary() {
  return (
    <>
      <Title muted level={3}>
        <LoremIpsum short />
      </Title>

      <Title inverted level={3}>
        <LoremIpsum short />
      </Title>

      <Title primary level={3}>
        <LoremIpsum short />
      </Title>
    </>
  );
}

withDifferentStatesMutedInvertedAndPrimary.story = {
  name: 'With different states: muted, inverted, and primary.',
};

export function withAlignedText() {
  return (
    <>
      <Title level={3}>
        <LoremIpsum short />
      </Title>

      <Title centerAlign level={3}>
        <LoremIpsum short />
      </Title>

      <Title endAlign level={3}>
        <LoremIpsum short />
      </Title>
    </>
  );
}

withAlignedText.story = {
  name: 'With aligned text.',
};
