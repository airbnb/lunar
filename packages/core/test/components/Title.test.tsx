import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Title from '../../src/components/Title';

describe('<Title />', () => {
  it('errors when multiple alignments are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Title level={1} centerAlign endAlign>
          Default
        </Title>,
      );
    }).toThrowError();
  });

  it('errors when multiple states are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Title level={1} muted inverted>
          Default
        </Title>,
      );
    }).toThrowError();
  });

  it('renders level 1', () => {
    const wrapper = shallowWithStyles(<Title level={1}>Title</Title>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders level 2', () => {
    const wrapper = shallowWithStyles(<Title level={2}>Title</Title>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders level 3', () => {
    const wrapper = shallowWithStyles(<Title level={3}>Title</Title>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inline', () => {
    const wrapper = shallowWithStyles(
      <Title level={1} inline>
        Inline
      </Title>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inverted', () => {
    const wrapper = shallowWithStyles(
      <Title level={1} inverted>
        Inverted
      </Title>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders muted', () => {
    const wrapper = shallowWithStyles(
      <Title level={1} muted>
        Muted
      </Title>,
    );

    expect(wrapper).toMatchSnapshot();
  });
});
