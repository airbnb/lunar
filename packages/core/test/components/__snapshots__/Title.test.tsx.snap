// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Title /> renders inline 1`] = `
<h1
  className="title_gfa0uf-o_O-title_level1_17a30zo-o_O-title_inline_g86r3e"
>
  Inline
</h1>
`;

exports[`<Title /> renders inverted 1`] = `
<h1
  className="title_gfa0uf-o_O-title_level1_17a30zo-o_O-title_inverted_bd4m75"
>
  Inverted
</h1>
`;

exports[`<Title /> renders level 1 1`] = `
<h1
  className="title_gfa0uf-o_O-title_level1_17a30zo"
>
  Title
</h1>
`;

exports[`<Title /> renders level 2 1`] = `
<h2
  className="title_gfa0uf-o_O-title_level2_1aurm9a"
>
  Title
</h2>
`;

exports[`<Title /> renders level 3 1`] = `
<h3
  className="title_gfa0uf-o_O-title_level3_1vnvwgc"
>
  Title
</h3>
`;

exports[`<Title /> renders muted 1`] = `
<h1
  className="title_gfa0uf-o_O-title_level1_17a30zo-o_O-title_muted_10k87om"
>
  Muted
</h1>
`;
